app/services/search/postgres/comment.rb

```diff
@@ -0,0 +1,89 @@
+module Search
+  module Postgres
+    class Comment
+      ATTRIBUTES = [
+        "articles.published",
+        "comments.body_markdown",
+        "comments.commentable_id AS commentable_id",
+        "comments.commentable_type AS commentable_type",
+        "comments.created_at",
+        "comments.id AS id",
+        "comments.public_reactions_count",
+        "comments.score",
+        "comments.user_id AS comment_user_id",
+        "podcast_episodes.podcast_id AS podcast_id",
+        "podcasts.published",
+        "users.id AS user_id",
+        "users.name",
+        "users.profile_image",
+        "users.username",
+      ].freeze
+      private_constant :ATTRIBUTES
+
+      DEFAULT_PER_PAGE = 60
+      private_constant :DEFAULT_PER_PAGE
+
+      # Because commentable represents a polymorphic relationship, ActiveRecord
+      # can't eager load the associations so we have to make all the joins
+      # manually.
```

Comment:
Comments are built using a polymorphic relationship (`commentable`). In working on this I realized that polymorphism introduces several...challenges(?) when trying to do PostgreSQL FTS. The way we handled this complexity with Elasticsearch is we built the documents beforehand, combining all the data we needed. We're not doing this with PostgreSQL so it gets a little complicated to get everything we need.

Technically, we could add more columns in the DB to make search easier and queries more efficient. I was trying to avoid that route if we could, but I'm open to it!

Reply:
Which columns would you add? Can we use the approach @rhymes used for articles (?), i.e. combining triggers and a `tsvector` column?

Reply:
@rhymes, addressing [your other comment](https://github.com/forem/forem/pull/13401#discussion_r613919746) here as well...

I only have 33 comments locally. 12k is solid, let's do it!

I have a few screenshots...admittedly, I haven't been including any rough data on performance [since @jgaskins pointed out that it's different in development either way](https://github.com/forem/forem/pull/12975#pullrequestreview-610866651). We don't have access to prod snapshots so I was hoping @jgaskins could help us again here. I'm open to running more tests/helping however I can!

```ruby
::Comment
  .joins(FORCED_EAGER_LOAD_QUERY)
  .where(QUERY_FILTER)
  .search_comments(term)
  .with_pg_search_highlight
  .analyze
```

**As is, no indexes added**
![Screen Shot 2021-04-14 at 12 32 35 PM](https://user-images.githubusercontent.com/15987080/114916472-d53b1e80-9df2-11eb-91d3-8c4ff07a074c.png)

**Added an index on `comments.deleted` and `comments.hidden_by_commentable_user`, separately**
![Screen Shot 2021-04-14 at 12 43 21 PM](https://user-images.githubusercontent.com/15987080/114920769-cacf5380-9df7-11eb-967a-8b31e759da6b.png)

**Added an index on `comments.deleted` and `comments.hidden_by_commentable_user`, together (removed previous indexes)**
![Screen Shot 2021-04-14 at 12 44 31 PM](https://user-images.githubusercontent.com/15987080/114920786-d0c53480-9df7-11eb-831f-48cba0be9da1.png)

@citizen428

> Can we use the approach @rhymes used for articles (?), i.e. combining triggers and a tsvector column?

I'm trying to avoid this as much as possible since it adds quite a bit of complexity. Since the primary goal is to remove Elasticsearch, I was hoping we can delay that approach for whenever we circle back around with a primary goal of improving search. At that time, we should also have more helpful insights on how search is being used, if at all.

Reply:
The difference between index and no index is negligeable, and I think we'll need them in production. That said, we can go about it in iterations. We can merge this approach sans the N + 1 queries, then we test it in production and regroup :) 

My hunch is that we'll need those indexes but it's better to talk with some hard data in front of our eyes.